### PR TITLE
Small refactoring of challenge-based protocols

### PR DESCRIPTION
**Pull request by georgwiese**
Some small refactorings of `lookup.asm` and `permutation.asm`.

### CODE REVIEW

File: std/protocols/lookup.asm [hunk 1]
```diff
@@ -67,11 +64,6 @@ let lookup: expr, expr[], Fp2<expr>, Fp2<expr>, Constr, expr -> Constr[] = |is_f
 
     let (lhs_selector, lhs, rhs_selector, rhs) = unpack_lookup_constraint(lookup_constraint);
 
-    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = if !is_extension(acc) {
-        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
-    } else { };
```
Comment (georgwiese) [inline std/protocols/lookup.asm:73]: Done inside `fp2_from_array` now.

File: std/protocols/lookup.asm [hunk 2]
```diff
@@ -80,13 +72,7 @@ let lookup: expr, expr[], Fp2<expr>, Fp2<expr>, Constr, expr -> Constr[] = |is_f
     let lhs_denom = sub_ext(beta, fingerprint(lhs, alpha));
     let rhs_denom = sub_ext(beta, fingerprint(rhs, alpha));
     let m_ext = from_base(multiplicities);
-
-    let next_acc = if is_extension(acc) {
-        next_ext(acc_ext)
-    } else {
-        // The second component is 0, but the next operator is not defined on it...
-        from_base(acc[0]')
-    };
```
Comment (georgwiese) [inline std/protocols/lookup.asm:89]: Seems like this has been fixed in the meantime :)

File: std/math/fp2.asm [hunk 3]
```diff
@@ -111,7 +116,16 @@ let is_extension = |arr| match len(arr) {
 };
 
 /// Constructs an extension field element `a0 + a1 * X` from either `[a0, a1]` or `[a0]` (setting `a1`to zero in that case)
-let fp2_from_array = |arr| if is_extension(arr) { Fp2::Fp2(arr[0], arr[1]) } else { from_base(arr[0]) };
+let fp2_from_array = |arr| {
+    let _ = if !is_extension(arr) {
```
Comment (chriseth) [inline std/math/fp2.asm:120]: I think you don't need `let _ = ` any more here.

File: std/math/fp2.asm [hunk 4]
```diff
@@ -111,7 +116,16 @@ let is_extension = |arr| match len(arr) {
...
+        assert(!needs_extension(), || "The field is too small and needs to move to the extension field. Pass two elements instead!")
+    } else { };
+    if is_extension(arr) {
```
Comment (chriseth) [inline std/math/fp2.asm:123]: Why not combine those?

File: std/protocols/lookup.asm [hunk 5]
```diff
@@ -67,11 +64,6 @@ let lookup: expr, expr[], Fp2<expr>, Fp2<expr>, Constr, expr -> Constr[] = |is_f
 
     let (lhs_selector, lhs, rhs_selector, rhs) = unpack_lookup_constraint(lookup_constraint);
 
-    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
```
[the remark this answers is not in the document]
Comment (chriseth) [reply std/protocols/lookup.asm:70]: But you can work around that by just using the enum manually without the `in` operator.

[the remark this answers is not in the document]
Comment (Schaeff) [reply std/protocols/lookup.asm:70]: The enum now stores an array of pairs so @georgwiese is right?

[the remark this answers is not in the document]
Comment (georgwiese) [reply std/protocols/lookup.asm:70]: I think @Schaeff is right ;) See #1440.

File: std/math/fp2.asm (hunk 3 above)
Comment (georgwiese) [reply std/math/fp2.asm]: Looks like we still do! I got `Expected constraint but got ()`. Maybe `assert` should return `[]`?

File: std/protocols/lookup.asm [hunk 6]
```diff
@@ -67,11 +64,6 @@ let lookup: expr, expr[], Fp2<expr>, Fp2<expr>, Constr, expr -> Constr[] = |is_f
...
-    let _ = if !is_extension(acc) {
-        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
-    } else { };
-
     // On the extension field, we'll need two field elements to represent the challenge.
     // If we don't need an extension field, we can simply set the second component to 0,
     // in which case the operations below effectively only operate on the first component.
```
Comment (onurinanc) [inline std/protocols/lookup.asm:69]: We can take this explanation from here to `fp2_from_array` inside fp2.asm

File: std/protocols/permutation.asm [hunk 7]
```diff
@@ -84,29 +81,17 @@ let permutation: expr, expr[], Fp2<expr>, Fp2<expr>, Constr -> Constr[] = |is_fi
...
     let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);
 
-    let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = if !is_extension(acc) {
-        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
-    } else { };
-
     // On the extension field, we'll need two field elements to represent the challenge.
     // If we don't need an extension field, we can simply set the second component to 0,
     // in which case the operations below effectively only operate on the first component.
```
Comment (onurinanc) [inline std/protocols/permutation.asm:86]: Same as in the lookup.asm, we can move this explanation to `fp2_from_array`, and delete this

File: std/math/fp2.asm [hunk 8]
```diff
@@ -111,7 +116,14 @@ let is_extension = |arr| match len(arr) {
...
 /// Constructs an extension field element `a0 + a1 * X` from either `[a0, a1]` or `[a0]` (setting `a1`to zero in that case)
-let fp2_from_array = |arr| if is_extension(arr) { Fp2::Fp2(arr[0], arr[1]) } else { from_base(arr[0]) };
+let fp2_from_array = |arr| {
+    if is_extension(arr) {
+        Fp2::Fp2(arr[0], arr[1])
+    } else {
+        let _ = assert(!needs_extension(), || "The field is too small and needs to move to the extension field. Pass two elements instead!");
+        from_base(arr[0])
+    }
+};
```
Comment (onurinanc) [inline std/math/fp2.asm:126]: This looks nice

**Review by onurinanc — APPROVED** [review]

**Pull request merged**